pkg/scheduler/actions/backfill/backfill.go

```diff
@@ -68,6 +74,101 @@ func (alloc *backfillAction) Execute(ssn *framework.Session) {
 			}
 		}
 	}
+
+	if len(alloc.arguments) == 0 {
+		return
+	}
+
+	backfillNonBestEffortJobsEnabled, err := strconv.ParseBool(alloc.arguments[conf.BackfillFlagName])
+	if err != nil {
+		glog.V(3).Infof("backfilling non best effort jobs is disabled")
+		return
+	}
+
+	if backfillNonBestEffortJobsEnabled {
+		// Collect back fill candidates
+		backFillCandidates := make([]*api.JobInfo, 0, len(ssn.Jobs))
+		for _, job := range ssn.Jobs {
+			if !ssn.BackFillEligible(job) {
+				continue
+			}
+			backFillCandidates = append(backFillCandidates, job)
+		}
+
+		// Release resources allocated to unready top dog jobs so that
+		// we can back fill more jobs in the next step.
+		resourceReleased := false
+		for _, job := range ssn.Jobs {
+			if ssn.JobReady(job) || ssn.JobBackfillReady(job) {
+				glog.V(3).Infof("Disable backfill on job <%v/%v>", job.Namespace, job.Name)
+			} else {
+				releaseReservedResources(ssn, job)
+				resourceReleased = true
+			}
+		}
+
+		if resourceReleased {
+			for _, job := range backFillCandidates {
+				backFill(ssn, job)
+			}
+		}
+	}
+}
+
+// Releases resources allocated to the given job back to the cluster.
+func releaseReservedResources(ssn *framework.Session, job *api.JobInfo) {
+	glog.V(3).Infof("Releasing resources allocated to job <%v/%v>", job.Namespace, job.Name)
+
+	for _, task := range job.Tasks {
+		if task.Status == api.Allocated || task.Status == api.Borrowing {
+			if err := job.UpdateTaskStatus(task, api.Pending); err != nil {
+				glog.Errorf("Failed to update task <%v/%v> status to %v in Session <%v>: %v",
+					task.Namespace, task.Name, api.Pending, ssn.UID, err)
+			}
+
+			node := ssn.Nodes[task.NodeName]
+			if err := node.RemoveTask(task); err != nil {
+				glog.Errorf("Failed to remove task %v from node %v: %s", task.Name, node.Name, err)
+				continue
+			}
+
+			glog.V(4).Infof("Removed task %s from node %s. Idle: %+v; Used: %v; Releasing: %v.", task.Name, node.Name, node.Idle, node.Used, node.Releasing)
+		}
+	}
+}
+
+func backFill(ssn *framework.Session, job *api.JobInfo) {
+	glog.V(3).Infof("Backfill job <%v/%v>", job.Namespace, job.Name)
+
+	for _, task := range job.TaskStatusIndex[api.Pending] {
+		allocateFailed := false
+		for _, node := range ssn.Nodes {
```

Comment:
how about nodeOrder?

Reply:
will be addressed along with the starvation PR

Reply:
I think we already sync up about this several times :( , move the fix into this PR.

Reply:
This is backfill. Backfill does not require stable node order. The node order does not matter here.

Reply:
we need to call `nodeOrderFn` to get the "best" resources for it.

Reply:
@k82cn  You are right. We should use `util.PredicateNodes`, `util.PrioritizeNodes`, and `util.SelectBestNode` as we do in allocate.go.

Reply:
> I think we already sync up about this several times :( , move the fix into this PR.

Same thing with starvation PR is based on the backfill PR which needs to be merged first. 

Same thing with backfill PR is a big PR and saying that it should be split into smaller PRs does not help. 

Same thing with we need to prioritize comments

Yeah, we have sync'd on these too so many times and look where we are now.